File: pset1 - C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool luhn_algorithm(long number);
int get_first_digits(long number);
int count_digits(long number);

int main(void)
{
    long number = get_long("Número: "); // Obtém o número do usuário

    // Verifica se o número é válido pelo Algoritmo de Luhn
    if (!luhn_algorithm(number))
    {
        printf("INVALID\n");
        return 0;
    }

    int length = count_digits(number);
    int first_digits = get_first_digits(number);

    // Verifica a bandeira do cartão
    if ((length == 15) && (first_digits == 34 || first_digits == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 16) && (first_digits >= 51 && first_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && (first_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Implementação do Algoritmo de Luhn
bool luhn_algorithm(long number)
{
    int sum = 0;
    int position = 0;

    while (number > 0)
    {
        int digit = number % 10;

        // Multiplica por 2 os dígitos em posição par (da direita para a esquerda)
        if (position % 2 == 1)
        {
            int doubled = digit * 2;
            sum += (doubled / 10) + (doubled % 10);
        }
        else
        {
            sum += digit;
        }

        number /= 10;
        position++;
    }

    return (sum % 10 == 0);
}

// Conta o número de dígitos do cartão
int count_digits(long number)
{
    int count = 0;
    while (number > 0)
    {
        number /= 10;
        count++;
    }
    return count;
}

// Obtém os dois primeiros dígitos do cartão
int get_first_digits(long number)
{
    while (number >= 100)
    {
        number /= 10;
    }
    return number;
}

File: pset6 - Python/credit/credit.py
from cs50 import get_string


def luhn_check(card_number):
    total = 0
    reverse_digits = card_number[::-1]

    for i, digit in enumerate(reverse_digits):
        n = int(digit)
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9
        total += n

    return total % 10 == 0


def get_card_type(card_number):
    length = len(card_number)
    first_two = int(card_number[:2])
    first_one = int(card_number[:1])

    if length == 15 and first_two in {34, 37}:
        return "AMEX"
    elif length == 16 and 51 <= first_two <= 55:
        return "MASTERCARD"
    elif length in {13, 16} and first_one == 4:
        return "VISA"
    else:
        return "INVALID"


card_number = get_string("Number: ")

if luhn_check(card_number):
    print(get_card_type(card_number))
else:
    print("INVALID")
